src/runtime/propensity-server-test.js

```diff
@@ -296,4 +312,132 @@ describes.realWin('PropensityServer', {}, env => {
           expect(() => {throw reason;}).to.throw(/Invalid request/);
         });
   });
+
+  it('should listen for events from event manager', function*() {
+    //event registration happens during initialization so we need to stub it
+    //and then create a new propensity server
+    const eventMan = new ClientEventManager();
+    let receivedEvent = null;
+    sandbox.stub(PropensityServer.prototype, 'sendClientEvent_', event => {
```

Comment:
Line 321 and similar:
Instead of stubbing an internal method of the class under test, the right way to do is to mock the method of EventManager, registerEventListener and capture the argument to that call. and that verifies that PropensityServer listens for events from the event manager. None of the event manager specific code should be tested here.

A helper method that returns the event sent in the URL
/**
 * @param {string} capturedUrl
 * @return {?PropensityEvent}
 */
 getPropensityEventFromUrl(capturedUrl) {
        const queryString = capturedUrl.split('?')[1];
        const queries = parseQueryString(queryString);
        expect('events' in queries).to.be.true;

        const events = decodeURIComponent(queries['events'].split(':'));
       const eventName = events[1];
       const eventParam = parseJson(events[2]);
       const propensityEvent = /** @typedef {PropensityEvent} */({
         name: eventName,
         active: eventParam['is_active'],
         data: eventParam,
       } 
        expect(events).to.equal(JSON.stringify(eventParam));

The best way to test the mapping is to capture what gets sent in the fetch API.
    let capturedUrl;
    let capturedRequest;
    sandbox.stub(Xhr.prototype, 'fetch',
        (url, init) => {
          capturedUrl = url;
          capturedRequest = init;
          return Promise.reject('Server down');
        });

Then when you invoke the captured handler with different events, you can test your can test what was sent in the URL
return capturedListenerCallback(defaultAnalyticsEvent).then(() => {
  throw new Error('must have failed');
}).catch(() => {
        const propensityEventSent = this.getPropensityEventFromUrl(capturedUrl);
       // expectation on this event.
        expect(() => {
          throw reason;
        }).to.throw(/Server down/);
});

Reply:
Done!